File: python/pipeline.py
class Pipeline:
    def __init__(self, config, emailer, log):
        self.config = config
        self.emailer = emailer
        self.log = log

    def run(self, project):
        tests_passed = False
        deploy_successful = False

        if project.has_tests():
            if "success" == project.run_tests():
                self.log.info("Tests passed")
                tests_passed = True
            else:
                self.log.error("Tests failed")
                tests_passed = False
        else:
            self.log.info("No tests")
            tests_passed = True

        if tests_passed:
            if "success" == project.deploy():
                self.log.info("Deployment successful")
                deploy_successful = True
            else:
                self.log.error("Deployment failed")
                deploy_successful = False
        else:
            deploy_successful = False

        if self.config.send_email_summary():
            self.log.info("Sending email")
            if tests_passed:
                if deploy_successful:
                    self.emailer.send("Deployment completed successfully")
                else:
                    self.emailer.send("Deployment failed")
            else:
                self.emailer.send("Tests failed")
        else:
            self.log.info("Email disabled")

File: typescript/src/Pipeline.ts
import {Config} from "./dependencies/Config";
import {Mailer} from "./dependencies/Mailer";
import {Logger} from "./dependencies/Logger";
import {Project} from "./dependencies/Project";

export class Pipeline {
    private readonly config: Config;
    private readonly mailer: Mailer;
    private readonly log: Logger;

    constructor(config: Config, mailer: Mailer, log: Logger) {
        this.config = config;
        this.mailer = mailer;
        this.log = log;
    }

    run(project: Project): void {
        let testsPassed: boolean;
        let deploySuccessful: boolean;

        if (project.hasTests()) {
            if ("success" === project.runTests()) {
                this.log.info("Tests passed");
                testsPassed = true;
            } else {
                this.log.error("Tests failed");
                testsPassed = false;
            }
        } else {
            this.log.info("No tests");
            testsPassed = true;
        }

        if (testsPassed) {
            if ("success" === project.deploy()) {
                this.log.info("Deployment successful");
                deploySuccessful = true;
            } else {
                this.log.error("Deployment failed");
                deploySuccessful = false;
            }
        } else {
            deploySuccessful = false;
        }

        if (this.config.sendEmailSummary()) {
            this.log.info("Sending email");
            if (testsPassed) {
                if (deploySuccessful) {
                    this.mailer.send("Deployment completed successfully");
                } else {
                    this.mailer.send("Deployment failed");
                }
            } else {
                this.mailer.send("Tests failed");
            }
        } else {
            this.log.info("Email disabled");
        }
    }
}
